src/main/java/com/capgemini/fs/coindashboard/apiCommunicator/ApiHolder.java

```diff
@@ -30,58 +27,37 @@ public ApiCommunicator getApiCommunicator(ApiProviderEnum apiProviderEnum) {
     return this.apiCommunicators.get(apiProviderEnum);
   }
 
-  public List<CoinMarketDataResult> getCoinMarketData(
+  public CoinMarketDataResult getCoinMarketData(
       String coinName) { // TODO: handle crypto names (ie. btc = bitcoin = 1 inCMC)
-    List<CoinMarketDataResult> results = new ArrayList<>();
-    Iterator<Entry<ApiProviderEnum, ApiCommunicator>> iterator = this.apiCommunicators.entrySet()
-        .iterator();
-    boolean stop = false;
-    while (iterator.hasNext() && !stop) {
-      Map.Entry<ApiProviderEnum, ApiCommunicator> entry = iterator.next();
-      if (entry.getKey() == ApiProviderEnum.COIN_GECKO && Objects.equals(coinName, "btc")) {
-        coinName = "bitcoin";
-      }
-      String finalCoinName = coinName;
-      CoinMarketDataResult result = entry.getValue().getCurrentListing(new ArrayList<>() {{
-        add(finalCoinName);
-      }}, new ArrayList<>() {{
-        add("usd");
-      }});
-      results.add(result);
-      if (result.getStatus() == ResultStatus.SUCCESS) {
-        stop = true;
-      }
-    }
-    return results;
+
+    return this.apiCommunicators.entrySet().stream()
+        .map(
+            api ->
+                api.getValue()
+                    .getCurrentListing(
+                        new ArrayList<>(
+                            List.of(
+                                api.getKey() == ApiProviderEnum.COIN_GECKO ? "bitcoin" : coinName)),
+                        new ArrayList<>(List.of("usd"))))
```

Comment:
Better have a good explanation!!! ;) 

As a bonus few words:

If we have a choice to use list that allows or disallows adding/removing elements without changing code logic (code works the same in both cases) then it's most often better to use list that disallows that. You are then secured against incorrect modification (like accidental removing element from list). 

The same with final or not final reference (accidental assigning to non-final variable), mutable or immutable object.

Reply:
Is "Honestly, I have no idea how java works" a good explanation? Because that would be the most accurate one. Before that I used new ArrayList(){{add(..);}}, which is bad, so I searched for a oneline collection initializer and that's what I found. I didn't think that java would booby trap a collection constructor, so I didn't look any further into that. I will try to fix this in the future, thank you.